### ISSUE TITLE
Data Loss on Disconnect with Python

### ISSUE THREAD

**Issue by TimPansino** [grpc/grpc#29110]
### What version of gRPC and what language are you using?

Python gRPC sync channels, dates back at least 4 years and up through at least commit ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b or version v1.44.0

### What operating system (Linux, Windows,...) and version?

Any, tested on Linux and Mac

### What runtime / compiler are you using (e.g. python version or version of gcc)

Python, any supported version. Tested on 3.8 and 3.9 recently.

### What did you do?

Using gRPC on a background thread that sends up data from a stream buffer. The buffer is iterable and passed as the request iterator. When calling `__next__` it is possible for the stream buffer to be empty, in which case the thread sleeps until a condition alerts the thread that another item has been added to the queue for transmission.

On the server side, the implementation issues a status code OK every 30 seconds that makes clients reconnect for load balancing. I'm aware this isn't exactly recommended.

### What did you expect to see?

Data consumed by the iterator should be properly sent to the server.

### What did you see instead?

The function [consume_request_iterator](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L159) contains a loss of data on disconnect.

When the status code changes to OK, the iterator that's currently stalled out while retrieving the next item will still consume the item (nothing can be done about that on the gRPC end). Unfortunately, that consumed item will be dropped and not sent based on the code in `consume_request_iterator`.

Specifically, [this line](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L178) checks the status code *after* the request iterator returns, and if the status code is set it will exit and silently drop the item it was supposed to transmit.

This means for our application data loss would occur on 1 item every 30 seconds, guaranteed. We've mitigated that in our code now with a custom iterator for our buffer that handles checking the status code on the channel, and will raise a stop iteration if it is set rather than return the next item. This will prevent the main issue we experienced, where every disconnect issued was causing data loss.

Unfortunately from investigating this we've noticed it is *still* possible to lose data. If the serialization of the request or acquiring the state's condition are taking place when the disconnect is issued there is still a problem. Both of these occur [right before checking the status code](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L176), but after consuming the next item from the request iterator. Which of course means data loss.

I'm not sure anything can be done about this, it seems like the dropped data is an intentional side effect of the design. Wanted to bring this up and get opinions from the maintainers in case this was simply overlooked or if any other languages have solved this problem.

### Anything else we should know about your project / environment?

Here are the relevant changes made for safeguarding our agent against this issue, if they are of interest to anyone reading this in the future for informational purposes.

https://github.com/newrelic/newrelic-python-agent/pull/490/files

**Comment by gnossen** [comment]
@TimPansino Thank you very much for the detailed write-up!

I want to make sure I'm understanding your question correctly, so let me reiterate it and please correct me if I'm getting any of this wrong. I think the situation in which you lose an event is as follows:

- Your StreamBuffer empties and it sleeps, waiting for some other thread to enqueue a new event. At this point, [the consumption thread](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L224) is sleeping [while calling next on the request iterator](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L165).
- The server sends an OK back to the client.
- The client receives the status and closes the connection, not drawing any more events from the request iterator (i.e. the StreamBuffer)
- Crucially, the [consumption thread](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L223) remains stuck [calling `next` on the request iterator](https://github.com/grpc/grpc/blob/ca7ba4d0ac3ab452c5db60befc8be37fd6e2339b/src/python/grpcio/grpc/_channel.py#L165) 
- (At some point, potentially much later) The StreamBuffer receives another event from a different thread, reviving the consumption thread.
- Since the stream has ended, the event is guaranteed to be dropped.

This does indeed sound like a problem. What's more, the library's hands are somewhat tied here. We do not have the ability to interrupt user code once entered. Ideally, we could provide you a condition to sleep on. But then you may have your own condition you need to sleep on. You really need some mechanism like [`select`](https://man7.org/linux/man-pages/man2/select.2.html) to wait on multiple conditions at once, but there doesn't seem to be anything like this in the 
Python threading API.

I think more thought/design is needed here. We're more than open to any suggestions you might have.

CC @lidizheng

**Comment by TimPansino** [comment]
That is correct, and I also came up empty handed looking for a solution from the library end. The only thing that came to mind was the ability for gRPC to take an optional condition in that it could acquire, allowing some more complex designs from the user side. It could allow the library to use a condition that's paired with the iterator, so that `__next__` isn't called until there's more in the queue.

I'm not certain that's the correct path to go down but it seemed interesting. We don't currently have the capacity to contribute a change of that scale either.

**Comment by lidizheng** [comment]
@TimPansino Is adopting AsyncIO an option here? In AsyncIO, we support `read`/`write` API for streaming RPCs. The `write` API won't continue unless the message has plumbed through Core and the write-successful event has populated. Here is an example:

```python

call = stub.FullDuplexCall()

# Writes two requests, and pending two requests
await call.write(msg)
await call.write(msg)
await call.done_writing()

# But read should be unaffected
response = await call.read()
handle(response)
response = await call.read()
handle(response)

code = await call.code()
```

The asynchronous generator approach is also supported, see [async route guide example](https://github.com/grpc/grpc/blob/2231c2ba77cf22f3c8c302d91209c1c3f2f0632f/examples/python/route_guide/asyncio_route_guide_client.py#L106).

---

That being said, AsyncIO can be quite intrusive to adopt, it requires updating function signatures to propagate async-ness of calls.
